### PR TITLE
Add CI workflow for performance benchmarks

### PR DESCRIPTION
**Pull request by EwoutH**
This PR introduces a new GitHub Actions workflow for performance benchmarking. The workflow is triggered on `pull_request_target` events and can be triggered with a "/rerun-benchmarks" comment in the PR. It should be compatible with PRs from both forks and the main repository. It includes steps for setting up the environment, checking out the PR and main branches, installing dependencies, and running benchmarks on both branches. The results are then compared, encoded, and posted as a comment on the PR.

It's largely based and supersedes https://github.com/projectmesa/mesa/pull/1978 and is reasonably well tested on my own fork in https://github.com/EwoutH/mesa/pull/22. It shouldn't need any additional permissions.

The results is a comment on a PR like this:

Screenshot_349

This PR will start working when it's merged onto the main branch.

### CODE REVIEW

**Review by rht — APPROVED** [review]

**Comment by rht** [comment]
I might have missed some detail, but how did you fix the permission problem, if you could summarize? This is for my learning purpose.

**Comment by jackiekazil** [comment]
Catching up - this looks like a great addition. Since I have been tracking less -- i am going to defer to others for the review and merge. Let me know if you need anything.

File: .github/workflows/benchmarks.yml
Comment (Corvince) [inline .github/workflows/benchmarks.yml:36]: You forgot to update this line

File: .github/workflows/benchmarks.yml
Comment (Corvince) [inline .github/workflows/benchmarks.yml:11]: Are these needed with pull_request_target?

**Review by Corvince — COMMENTED** [review]
So happy you got it working! Added two comments

File: .github/workflows/benchmarks.yml
Comment (EwoutH) [reply .github/workflows/benchmarks.yml:11]: I think these are the default values, but explicit might be better than implicit here, especially if they ever change the defaults.

Comment (tpike3) [reply .github/workflows/benchmarks.yml:11]: This does make me a little nervous with "write" permissions. [Did you try it with "none" or "read" ](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs)  or is it even necessary, looking through some other projects with `.github/workflows/benchmarks.yml` I did not see others with permissions but it was a cursory look and there are many variants of repo setups and workflows.

Comment (EwoutH) [reply .github/workflows/benchmarks.yml:11]: > [Did you try it with "none" or "read" ](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs)or is it even necessary,

Yes, doesn't work. Basically the whole discussion of Friday: https://github.com/projectmesa/mesa/pull/1978#issuecomment-1901104588. This is because there is write permission needed to place a comment.

**Comment by EwoutH** [comment]
> I might have missed some detail, but how did you fix the permission problem, if you could summarize? This is for my learning purpose.

Using the [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target). That does have write permissions, but only initiated from the main branch of a repo. So it only takes effect once it's merged into main, protecting it from attacks on CI infrastructure.

File: .github/workflows/benchmarks.yml
Comment (EwoutH) [reply .github/workflows/benchmarks.yml]: Good catch, done

**Comment by EwoutH** [comment]
So let's see if this works: Let's see if the CI is working: /rerun-benchmarks

**Comment by EwoutH** [comment]
It works! Still a bit noisy, but enough to get a general picture.

For future improvements we could:
- Ensure seeds are correctly working
- Run on a own, custom runner without hyperthreading, clock speed turbo boosting and conflicting tasks
- Optimize configurations, replications and statistical methods
- Add models that test relevant components

**Comment by EwoutH** [comment]
Note to self: It looks like the comment-triggerd workflows can't find the PR branch correctly. Needs a fix.

**Pull request merged**
